### ISSUE TITLE
Addition of native functions to match JavaScript counterparts

### ISSUE THREAD

**Issue by brandonmcconnell** [sass/sass#3205]
I have implemented several JavaScript core functions/methods in Sass/SCSS which has been very useful when working with different types and has actually made working with Sass even simpler for me as it feels just like JavaScript, another web language most Sass developers are familiar with.

I think it would be worthwhile to add these to Sass/SCSS core for the same reason. Including the few I've already made, my proposal would be to extend all the Sass built-in types with methods to make them as powerful—and as familiar—as their JavaScript counterparts.

These (JavaScript methods) include:
* `String.repeat()` as `str-repeat()`
* `String.padStart()` as `pad-start()`
* `String.padEnd()` as `pad-end()`
* `Array.flat()` as `list-flat()`
* `Array.map()` as `list-map()`
* `Array.flatMap()` as `list-flat-map()`
* `Array.filter()` as `list-filter()`
* `Array.reduce()` as `list-filter()`
* `Array.sort()` as `list-sort()`
* `Array.some()` as `list-some()`
* `Array.every()` as `list-every()`
* `!Array.some()` as `list-none()`
* `!Array.every()` as `list-not-every()`
* `Array.copyWithin()` as `list-copy-within()`
* `Array.find()` as `list-find()`
* `Array.reverse().find()` as `list-find-last()`
* `Array.findIndex()` as `list-find-index()`
* `Array.reverse().findIndex()` as `list-find-last-index()`
* `Array.indexOf()` as `list-index-of()`
* `Array.lastIndexOf()` as `list-last-index-of()`
* `Array.includes()` as `list-includes()`
* `!Array.includes()` as `list-not-includes()`
* `Array.pop()` as `list-unappend()`
* `Array.unshift()` as `list-prepend()`
* `Array.shift()` as `list-unprepend()`

I would be happy to help build and implement each of these to extend Sass's built-in types if there is an opportunity for me to do so. Otherwise, please accept this as a feature request.

# Specification Examples

**str-repeat**
```scss
@function str-repeat($repeat-string, $repeat-length) {
    $string: "";
    @for $i from 0 to $repeat-length {
        $string: $string + $repeat-string;
    }
    @return $string;
}
```

**pad-start**
```scss
@function pad-start($original-string, $target-length, $pad-string: " ") {
    $original-string: stringify($original-string);
    $pad-string: stringify($pad-string);
    @return str-slice(str-repeat($pad-string, $target-length) + $original-string, -1 * max($target-length, str-length($original-string)));
}
```

**pad-end**
```scss
@function pad-end($original-string, $target-length, $pad-string: " ") {
    $original-string: stringify($original-string);
    $pad-string: stringify($pad-string);
    @return str-slice($original-string + str-repeat($pad-string, $target-length), 0, max($target-length, str-length($original-string)));
}
```

# Example Use Cases

## Leading zeroes
This logic precedes each article's content with a 4-digit counter, with leading [stringified] zeroes for lower numbers, so in a list of 1,000 items, each of these strings below would be rendered based on nth-child (omitting all the in-between values):

```html
#0001. Article title
<!-- items 0002–0009 omitted for conciseness -->
#0010. Article title
<!-- items 0011–0099 omitted for conciseness -->
#0100. Article title
<!-- items 0101–0999 omitted for conciseness -->
#1000. Article title
```

```scss
main > article {
    @for $i from 1 through 1000 {
        &:nth-child(#{$i})::before {
            content: "##{pad-start($i, 4, 0)}. ";
        }
    }
}
```

## "Table of Contents" formatting
Sometimes when working with monospace fonts, it's helpful to use a character such as a space, period, or dash to separate left-side and right-side strings. This is common practice in tables of contents and in restaurant menus. Here is an example:

```html
Entree

Chicken Dinner .......... $24.99
Garlic Butter Salmon .... $19.39
Ratatouille (vegan) ..... $17.59


Appetizer

Pea Soup ................. $4.79
Potatoes au Gratin ....... $7.89
```

The below example will render this exact output (above) in the content property, making use of all `stringify`, `pad-start`, `pad-end`, and `str-repeat`:

```scss
$menu: (
    "Entree": (
        "Chicken Dinner": 24.99,
        "Garlic Butter Salmon": 19.39,
        "Ratatouille (vegan)": 17.59
    ),
    "Appetizer": (
        "Pea Soup": 4.79,
        "Potatoes au Gratin": 7.89
    )
);
$longest-item-length: 0;
$longest-price-length: 0;
@each $category, $items in $menu {
    @each $item, $price in $items {
        $longest-item-length: max($longest-item-length, str-length($item) + 1);
        $longest-price-length: max($longest-price-length, str-length(stringify($price)) + 2);
    }
}
$character-padding-char: ".";
$character-padding: 2;
$br: "\A";
$menu-text: "";
$category-counter: 1;
@each $category, $items in $menu {
    $menu-text: $menu-text + $category + str-repeat($br, 2);
    $item-counter: 1;
    @each $item, $price in $items {
        $menu-text: $menu-text + pad-end($item + " ", $longest-item-length + $character-padding, $character-padding-char) + pad-start(" $" + $price, $longest-price-length + $character-padding, $character-padding-char);
        @if $item-counter != length($items) {
            $menu-text: $menu-text + $br;
            $item-counter: $item-counter + 1;
        }
    }
    @if $category-counter != length($menu) {
        $menu-text: $menu-text + str-repeat($br, 3);
        $category-counter: $category-counter + 1;
    }
}
body {
    font-family: consolas, monospace;
    white-space: pre;
    &::before {
        content: $menu-text;
    }
}
```

**Comment by jathak** [comment]
At this point, any new functions we add will be in the one of the built-in modules, not the global namespace (to avoid conflicting with new CSS functions), so any of these functions we add would be in forms like `list.map`, not `list-map`.

I've opened #3213 to track several of the higher-order list functions you mention, but supporting them is blocked on supporting anonymous functions, which is a much more involved new feature, so we don't currently have any plans to work on it.

`list.sort` is already tracked in #155.

None of the other functions mentioned here are useful enough in a CSS context to justify the added complexity of including them in the language directly, since they should all be relatively straightforward to implement as user-defined functions if necessary.

**Comment by brandonmcconnell** [comment]
@jathak thanks! That sounds like a good path to take here.

If I wanted to contribute and build some of these out open-source style, what's the best process to take in doing so? Can we could additional functions using SCSS-style logic, or must they be built in Ruby?

**Comment by jathak** [comment]
Sorry, when I said "implement as user-defined functions", I meant that users should implement them as needed for their own project. You could also release a third-party library of Sass functions for others to use, but it wouldn't be part of / endorsed by Sass itself.

Ruby wouldn't be involved at all. The original Sass compiler was written in Ruby, but it was deprecated years ago. The current reference implementation is [Dart Sass](https://github.com/sass/dart-sass), which supports [user-defined custom functions in JS](https://sass-lang.com/documentation/js-api/modules#CustomFunction) if you're using the JS compiled version on NPM (Dart custom functions are also supported via the Dart API, but most people don't use that unless their project is already implemented in Dart).

**Comment by brandonmcconnell** [comment]
@jathak Hi, that makes sense to me. I've implemented most of these already in user-defined functions. However, these are functions that are available in almost every programming language natively, which was my reason for opening this ticket initially. Lacking these functions, Sass is still powerful but limited compared to most programming standards.

**Comment by roydukkey** [comment]
@brandonmcconnell Just want to point you to toward https://github.com/roydukkey/sass-fairy if you're interested in having these in other projects.

**Comment by brandonmcconnell** [comment]
@roydukkey Thanks, Trent! This is very close to what I would have imagined if using an outside library. Looking at Sass Fairy's [`list.map`](https://sass-fairy.com/api/list/map) function, is `sass:meta` essentially being used as a workaround to achieving anonymous functions?

**Comment by roydukkey** [comment]
Not really. `meta.get-function()` is the only way to get function references in Sass, so they can be passed around and [invoked from other locations](https://sass-lang.com/documentation/modules/meta#call) (which happens inside [`list.map()`](https://github.com/roydukkey/sass-fairy/blob/ad7c878f596cfb755cbf79e6d4b1402ffb2b0855/packages/list/src/_map.sass#L44)).

You might also look at Sass Fairy's [reference functions](https://sass-fairy.com/docs/reference-functions) as they'll help to expose functions so consumers don't have to use `sass:meta`.

If you have anymore questions, let's move the conversation to the sass-fairy repo as to not pollute this repo further. :)
